### PR TITLE
Enable nhwc and constant folding in ci

### PR DESCRIPTION
**Pull request by BBuf**

### CODE REVIEW

File: oneflow/ir/lib/OneFlow/OneFlowOps.cpp [hunk 1]
```diff
@@ -484,7 +484,11 @@ llvm::SmallVector<Value, 4> BiasAddOp::NchwToNhwc(llvm::SmallVector<Value, 4> va
   return results;
 }
 
-bool NormalizationOp::IsNCHW() { return this->axisAttr().getValue().getSExtValue() == 1; }
+bool NormalizationOp::IsNCHW() {
```
Comment (BBuf) [inline oneflow/ir/lib/OneFlow/OneFlowOps.cpp:487]: fix oneflow/python/oneflow/test/graph/test_free_tensor_not_in_job.py bug when enable auto_nhwc in ci.

File: oneflow/ir/lib/OneFlow/Passes.cpp [hunk 2]
```diff
@@ -621,6 +621,10 @@ struct AutoNhwcPattern : public OpInterfaceRewritePattern<NCHWCompatible> {
 
  public:
   LogicalResult matchAndRewrite(NCHWCompatible op, PatternRewriter& rewriter) const override {
+    if (op->hasTrait<OpTrait::IsOpConfCompatible>()) {
```
Comment (BBuf) [inline oneflow/ir/lib/OneFlow/Passes.cpp:624]: auto nhwc pass exist only in gpu mode

File: oneflow/ir/lib/OneFlow/Passes.cpp [hunk 3]
```diff
@@ -671,17 +678,23 @@ struct AutoNhwcPattern : public OpInterfaceRewritePattern<NCHWCompatible> {
...
 
-bool IsRedundantTransposeMatch(ArrayAttr pre, ArrayAttr afe) {
-  const auto prePerm = pre.getValue();
-  const auto afePerm = afe.getValue();
+bool IsRedundantTransposeMatch(ArrayAttr pre, ArrayAttr afe, mlir::PatternRewriter& rewriter) {
+  const auto prePerm = pre.getValue().vec();
+  const auto afePerm = afe.getValue().vec();
   if (prePerm.size() == 4 && afePerm.size() == 4) {
     // handle nchw->nhwc->nchw: (0, 2, 3, 1) -> (0, 3, 1, 2)
     if (prePerm[0] == afePerm[0] && prePerm[1] == afePerm[3] && prePerm[2] == afePerm[1]
```
Comment (BBuf) [inline oneflow/ir/lib/OneFlow/Passes.cpp:686]: fix bug，this check func has failed in enisum tests.

**Review by mosout — APPROVED** [review]

**Comment by BBuf** [comment]
image

File: oneflow/ir/lib/OneFlow/Passes.cpp [hunk 4]
```diff
@@ -621,6 +621,13 @@ struct AutoNhwcPattern : public OpInterfaceRewritePattern<NCHWCompatible> {
 
  public:
   LogicalResult matchAndRewrite(NCHWCompatible op, PatternRewriter& rewriter) const override {
+    if (op->hasTrait<OpTrait::IsOpConfCompatible>()) {
+      oneflow::UserOpAdaptor op_adaptor(op->getOperands(), op->getAttrDictionary());
```
Comment (jackalcooper) [inline oneflow/ir/lib/OneFlow/Passes.cpp:625]: 后来发现用属性不一定匹配的 adaptor 在新版本可能有bug，现在都是用traits里面定义的函数比如
```
IsOpConfCompatible<void>::getDeviceTag(b)
```

Comment (BBuf) [reply oneflow/ir/lib/OneFlow/Passes.cpp]: done.

**Pull request merged**